File: flutter/craftd_module/lib/presentation/ui/button/craftd_button_builder.dart
import 'package:flutter/src/widgets/framework.dart';
import '../../../data/model/action/action_properties.dart';
import '../../../data/model/button/button_properties.dart';
import '../../builder/craftd_builder.dart';
import '../../craftd_view_listener.dart';
import 'craftd_button.dart';

class CraftDButtonBuilder extends CraftDBuilder<ButtonProperties> {
  CraftDButtonBuilder() : super(key: key);

  @override
  Widget craft(ButtonProperties model, CraftDViewListener listener) {
    return CraftDButton(
      buttonProperties: model,
      callback: () {
        if (model.actionProperties != null) {
          listener(model.actionProperties!);
        }
      },
    );
  }

  @override
  ButtonProperties fromJson(properties) {
    return ButtonProperties(
        text: properties["text"],
        textColorHex: properties["textColorHex"],
        textSize: properties["textSize"],
        textAllCaps: properties["textAllCaps"],
        fillMaxSize: properties["fillMaxSize"],
        backgroundHex: properties["backgroundHex"],
        actionProperties: ActionProperties.fromJson(
            properties[ActionProperties.key]));
  }

  static String key = "CraftDButton";
}

File: android_kmp/craftd-compose/src/commonMain/kotlin/com/github/codandotv/craftd/compose/ui/button/CraftDButtonBuilder.kt
package com.github.codandotv.craftd.compose.ui.button

import androidx.compose.runtime.Composable
import com.github.codandotv.craftd.androidcore.data.convertToElement
import com.github.codandotv.craftd.androidcore.data.model.base.SimpleProperties
import com.github.codandotv.craftd.androidcore.data.model.button.ButtonProperties
import com.github.codandotv.craftd.androidcore.presentation.CraftDComponentKey
import com.github.codandotv.craftd.androidcore.presentation.CraftDViewListener
import com.github.codandotv.craftd.compose.builder.CraftDBuilder

class CraftDButtonBuilder(
    override val key: String = CraftDComponentKey.BUTTON_COMPONENT.key
) : CraftDBuilder {
    @Composable
    override fun craft(model: SimpleProperties, listener: CraftDViewListener) {
        val buttonProperties = model.value.convertToElement<ButtonProperties>()
        buttonProperties?.let {
            CraftDButton(it) {
                buttonProperties.actionProperties?.let { listener.invoke(it) }
            }
        }
    }
}